hashring/rbtree.go

```diff
@@ -234,55 +247,88 @@ func (t *redBlackTree) Delete(val int) bool {
 	return found != nil
 }
 
-func (n *redBlackNode) search(val int) (string, bool) {
-	if n.val == val {
-		return n.str, true
-	} else if val < n.val {
+func (n *redBlackNode) search(key keytype) (valuetype, bool) {
+	cmp := n.key.Compare(key)
+	if cmp == 0 {
+		return n.value, true
+	} else if 0 < cmp {
 		if n.left != nil {
-			return n.left.search(val)
+			return n.left.search(key)
 		}
 	} else if n.right != nil {
-		return n.right.search(val)
+		return n.right.search(key)
+	}
+	return nil, false
+}
+
+// walk Walks the tree in sorted order invoking `walker` for each node. If the
+// walker function returns `false`, walking is stopped and no more nodes will be
+// visited. Returns `true` if all nodes are visited; `false` if not.
+func (n *redBlackNode) walk(walker func(*redBlackNode) bool) bool {
```

Comment:
I think it does cover the behaviour. "returns true when the conditions is true for every node."

This is also why I think the one liner reads better

```
// returns true when the condition holds for all nodes on the left, this node, and all nodes on the right
return n.left.All(condition) && condition(n) && n.right.All(condition)
```